compressors/arithmetic_coding.py

```diff
@@ -0,0 +1,436 @@
+            if bit:
+                state += 1 << (self.params.PRECISION - num_bits_consumed - 1)
+            num_bits_consumed += 1
+        num_bits_consumed = self.params.PRECISION
+
+        # main decoding loop
+        while True:
+            # decode the next symbol
+            s = self.decode_step_core(low, high, state, self.freq_model.freqs_current)
+            low, high = ArithmeticEncoder.shrink_range(self.freq_model.freqs_current, s, low, high)
+            decoded_data_list.append(s)
+
+            # update the freq_model
+            self.freq_model.update_model(s)
+
+            # break when we have decoded all the symbols in the data block
+            if len(decoded_data_list) == input_data_block_size:
+                break
+
+            while (high < self.params.HALF) or (low > self.params.HALF):
+                if high < self.params.HALF:
+                    # re-adjust range, and reset params
+                    low = low << 1
+                    high = high << 1
+                    state = state << 1
+
+                elif low > self.params.HALF:
+                    # re-adjust range, and reset params
+                    low = (low - self.params.HALF) << 1
+                    high = (high - self.params.HALF) << 1
+                    state = (state - self.params.HALF) << 1
+
+                if num_bits_consumed < arith_bitarray_size:
+                    bit = encoded_bitarray[num_bits_consumed]
+                    state += bit
+                num_bits_consumed += 1
+
+            while (low > self.params.QTR) and (high < 3 * self.params.QTR):
+                # increment the mid-range adjustment counter
+                low = (low - self.params.QTR) << 1
+                high = (high - self.params.QTR) << 1
+                state = (state - self.params.QTR) << 1
+
+                if num_bits_consumed < arith_bitarray_size:
+                    bit = encoded_bitarray[num_bits_consumed]
+                    state += bit
+                num_bits_consumed += 1
+
+        # # NOTE: we might have loaded in additional bits not added by the arithmetic encoder
+        # # (which are present in the encoded_bitarray).
+        # # This block of code determines the extra bits and subtracts it from num_bits_consumed
+        for extra_bits_read in range(self.params.PRECISION):
+            state_low = (state >> extra_bits_read) << extra_bits_read
+            state_high = state_low + (1 << extra_bits_read)
+            if (state_low < low) or (state_high > high):
+                break
+        num_bits_consumed -= extra_bits_read - 1
+
+        # add back the bits corresponding to the num elements
+        num_bits_consumed += self.params.DATA_BLOCK_SIZE_BITS
+
+        return DataBlock(decoded_data_list), num_bits_consumed
+
+
+def test_bitarray_for_specific_input():
+    """
+    manually perform the encoding and check if the bitarray matches
+    """
+    # TODO: Kedar
```

Comment:
@Kedar: are you planning to implement it before accepting PR?

Reply:
yes ll add